webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/tracings/FossilDBClient.scala

```diff
@@ -184,7 +207,33 @@ class FossilDBClient(collection: String,
         case Empty    => Fox.empty
         case net.liftweb.common.Failure(msg, _, _) =>
           slackNotificationService.reportFossilWriteError("put", msg)
-          Fox.failure("could not save to FossilDB: " + msg)
+          Fox.failure("Could not save to FossilDB: " + msg)
+      }
+    } yield ()
+  }
+
+  def putMultiple(keyValueTuple: Seq[(String, Array[Byte])], version: Long, batchSize: Int = 1000): Fox[Unit] =
+    for {
+      _ <- Fox.serialCombined(keyValueTuple.grouped(batchSize))(batch => putMultipleImpl(batch, version))
+    } yield ()
+
+  private def putMultipleImpl(keyValueTuples: Seq[(String, Array[Byte])], version: Long): Fox[Unit] = {
+    val keyValuePairs = keyValueTuples.map {
+      case (key, value) => VersionedKeyValuePairProto(key, version, ByteString.copyFrom(value))
+    }
+    val putFox = for {
+      reply <- wrapException(
+        stub.putMultipleKeysWithMultipleVersions(PutMultipleKeysWithMultipleVersionsRequest(collection, keyValuePairs)))
+      _ <- assertSuccess(reply.success, reply.errorMessage)
+    } yield ()
+    for {
```

Comment:
Why start a second for here? can't this be avoided / shortened by leaving this out  and putting all in a single for comprehension?

Reply:
I still want to continue to the lower for comprehension if the upper one fails. Just continuing the same for comprehension would mean that the failure is returned, but the slack logging never happens.